File: problem-solving-basic/subarray-sums/sol.js
'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}



/*
 * Complete the 'findSum' function below.
 *
 * The function is expected to return a LONG_INTEGER_ARRAY.
 * The function accepts following parameters:
 *  1. INTEGER_ARRAY numbers
 *  2. 2D_INTEGER_ARRAY queries
 */

function findSum(numbers, queries) {
    // Write your code here
    let result=[];
    let subArraySum=[];
    let countZero=numbers[0]  ==  0  ?  1  :  0;
    let zeroArr=[];
    zeroArr[0]=countZero;
    subArraySum[0]=numbers[0];
    for (let i=1; i<=numbers.length-1;i++){
        if (numbers[i]==0){
            countZero++;
            zeroArr[i]=countZero;
        }
        else{
            zeroArr[i]=countZero;
        }
        subArraySum[i]=subArraySum[i-1]+numbers[i];
    }

    for (let q of queries){
        if (q.length==3)  {            
        let i=q[0];            
        let j=q[1];            
        let x=q[2];            
        let sum=0;                       
        sum=subArraySum[j-1] - ((i-2<0 ) ? 0 : subArraySum[i-2]);
        sum=sum+  (zeroArr[j - 1] - ((i-2<0) ? 0 : zeroArr[i - 2])) * x;
        result.push(sum);
        }  
    }
    return result;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const numbersCount = parseInt(readLine().trim(), 10);

    let numbers = [];

    for (let i = 0; i < numbersCount; i++) {
        const numbersItem = parseInt(readLine().trim(), 10);
        numbers.push(numbersItem);
    }

    const queriesRows = parseInt(readLine().trim(), 10);

    const queriesColumns = parseInt(readLine().trim(), 10);

    let queries = Array(queriesRows);

    for (let i = 0; i < queriesRows; i++) {
        queries[i] = readLine().replace(/\s+$/g, '').split(' ').map(queriesTemp => parseInt(queriesTemp, 10));
    }

    const result = findSum(numbers, queries);

    ws.write(result.join('\n') + '\n');

    ws.end();
}

File: problem-solving-basic/subarray-sums/sol.py
#!/bin/python3

import math
import os
import random
import re
import sys


#
# Complete the 'findSum' function below.
#
# The function is expected to return a LONG_INTEGER_ARRAY.
# The function accepts following parameters:
#  1. INTEGER_ARRAY numbers
#  2. 2D_INTEGER_ARRAY queries
#

def solution(numbers, queries):
    if not isinstance(numbers, list) and not isinstance(queries, list):
        return
    result = []
    total = 0
    for query in queries:
        if len(query) == 3:
            start = query[0]
            end = query[1]
            additional = query[2]
            selection = numbers[start-1:end]
            has_zeros = False
            for el in selection:
                if el == 0:
                    has_zeros = True
            total = sum(selection)
            if has_zeros:
                total += additional
            result.append(total)

    return result

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    numbers_count = int(input().strip())

    numbers = []

    for _ in range(numbers_count):
        numbers_item = int(input().strip())
        numbers.append(numbers_item)

    queries_rows = int(input().strip())
    queries_columns = int(input().strip())

    queries = []

    for _ in range(queries_rows):
        queries.append(list(map(int, input().rstrip().split())))

    result = findSum(numbers, queries)

    fptr.write('\n'.join(map(str, result)))
    fptr.write('\n')

    fptr.close()
